Schedule dynamic filtering collecting task immediately

sopel39 opened the pull request:
In case of plan
        J1
       /  \
     J2    S3
    /  \
   S1  S2

It might happen that that dynamic filtering dependencies are:
S3 => S3 => S1

With phased scheduler source stage consisting of
(J1, J2, S1) won't be scheduler until stages with S3 and S2
have finished split enumaration. However, it might happen
that S2 is waiting for dynamic filters for S3. In that case,
S3 will never complete because DFs for S3 are collected in
stage (J1, J2, S1).

This commit makes scheduling of DF collecting task immediately
which will prevent queries from deadlock.

raunaqmorarka commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveDynamicPartitionPruningTest.java:
```diff
@@ -37,6 +47,62 @@ protected QueryRunner createQueryRunner()
+        // t2 table scan depends on t1 DFs, but t0 <-> t2 join is blocked on t2 data
+        assertDynamicFilters(
+                noJoinReordering(joinDistributionType),
+                "SELECT v0, v1, v2 FROM (t0 JOIN t2 ON (k0 * -1) + 2 = i2) JOIN t1 ON k2 = i1");
```
It would be good to have a comment here that the expression `(k0 * -1) + 2 = i2` is being used instead of `k0 = i2` to avoid `t0` table scan having dependency on a DF

sopel39 commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/TestHiveDynamicPartitionPruningTest.java:
```diff
@@ -37,6 +47,62 @@ protected QueryRunner createQueryRunner()
     }
 
+    @Test(timeOut = 30_000, dataProvider = "joinDistributionTypes")
+    public void testJoinDynamicFilteringMultiJoin(FeaturesConfig.JoinDistributionType joinDistributionType)
```
I've added `createPartitionedTable` abstract method.

dain commented on plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergDynamicPartitionPruningTest.java:
```diff
@@ -38,7 +38,7 @@ protected QueryRunner createQueryRunner()
     protected void createLineitemTable(String tableName, List<String> columns, List<String> partitionColumns)
     {
-        String sql = format(
+        @Language("SQL") String sql = format(
```
There is a typo in the commit message

dain commented on core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:
```diff
@@ -541,13 +540,13 @@ public void close()
         checkState(allNodes.size() > 0, "No nodes available");
         InternalNode node = allNodes.get(ThreadLocalRandom.current().nextInt(0, allNodes.size()));
-        return scheduleTask(node, ImmutableMultimap.of(), ImmutableMultimap.of());
+        scheduleTask(node, ImmutableMultimap.of(), ImmutableMultimap.of());
```
In the old code this would return `Optional<RemoteTask>`.  Do we not need to handle the case where we got `Optional.empty()`?

sopel39 replied:
In old code this was used to return new tasks in `io.trino.execution.scheduler.ScheduleResult`. However, `io.trino.execution.scheduler.ScheduleResult#getNewTasks` is only used for testing purposes

dain commented on core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:
```diff
@@ -237,6 +237,12 @@ public synchronized void startLifespan(Lifespan lifespan, ConnectorPartitionHand
+        // schedule a task for collecting dynamic filters in case some dependent split generator is waiting for them
+        if (dynamicFilterService.isCollectingTaskNeeded(stageExecution.getStageId().getQueryId(), stageExecution.getFragment())) {
+            checkArgument(lifespan.isTaskWide(), "Dynamic filtering cannot be used with grouped execution");
+            createTaskOnRandomNode();
```
The old code as adding this task to `overallNewTasks`.  Do we not need to track this anymore?

sopel39 replied:
no, `io.trino.execution.scheduler.ScheduleResult#getNewTasks` is only used in tests.

dain commented on core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:
```diff
@@ -237,6 +237,12 @@ public synchronized void startLifespan(Lifespan lifespan, ConnectorPartitionHand
         whenFinishedOrNewLifespanAdded.set(null);
         whenFinishedOrNewLifespanAdded = SettableFuture.create();
+
+        // schedule a task for collecting dynamic filters in case some dependent split generator is waiting for them
```
I suggest we are more direct with this statement.  Something like
```
Avoid deadlocks by immediately scheduling a task for collecting dynamic filters, because there can be task in other stages blocked waiting for the dynamic filters.
```

dain commented on core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:
```diff
@@ -237,6 +237,12 @@ public synchronized void startLifespan(Lifespan lifespan, ConnectorPartitionHand
+
+        // schedule a task for collecting dynamic filters in case some dependent split generator is waiting for them
+        if (dynamicFilterService.isCollectingTaskNeeded(stageExecution.getStageId().getQueryId(), stageExecution.getFragment())) {
+            checkArgument(lifespan.isTaskWide(), "Dynamic filtering cannot be used with grouped execution");
```
If this is a new restriction, make sure to document it.

sopel39 replied:
It's not new. We have one in `io/trino/sql/planner/LocalExecutionPlanner.java:2871` already

dain commented on core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java:
```diff
@@ -232,7 +232,14 @@ public boolean isCollectingTaskNeeded(QueryId queryId, PlanFragment plan)
+        // dynamic filters produced within source stage are required by current stage to progress
+        // either directly (e.g. consumed by table scan) or indirectly (e.g. are required by join build
+        // source stage in order to complete).
+        return !getReplicatedDynamicFilters(plan.getRoot()).isEmpty();
```
I don't understand this change at all.  Someone else should review it.

joshthoward replied:
@raunaqmorarka can you describe this in more detail so that we can validate that this change is okay tomorrow?

sopel39 replied:
The change essentially does: If this is a source stage and some dynamic filters is produced by this stage (for broadcast join), then we need extra DF collecting task.

I've updated comment here:
```
            // Dynamic filters can be collected by single task only for source distribution (non-fixed) stage.
            // Such stage must only contain replicated joins.
```

dain reviewed (commented):
The scheduler stuff and the tests look good to me.  I don't understand the implications of the change to `DynamicFilterService`.

sopel39 commented:
ac @raunaqmorarka @dain . I've narrowed down when phased scheduler won't start join stage immediately to only non-fixed source stages. I've also improved tests with regards to source stage partitioning

dain commented on core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java:
```diff
@@ -490,10 +496,20 @@ private void collectDynamicFilters(StageId stageId, Optional<Set<DynamicFilterId
-        Set<DynamicFilterId> innerStageDynamicFilters = intersection(getProducedDynamicFilters(planNode), getConsumedDynamicFilters(planNode));
-        Set<DynamicFilterId> replicatedDynamicFilters = getReplicatedDynamicFilters(planNode);
-        return ImmutableSet.copyOf(intersection(innerStageDynamicFilters, replicatedDynamicFilters));
+        Set<DynamicFilterId> innerStageDynamicFilters = intersection(
```
you could inline this

dain reviewed (approved):
looks good

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/scheduler/SourcePartitionedScheduler.java:
```diff
@@ -237,6 +237,14 @@ public synchronized void startLifespan(Lifespan lifespan, ConnectorPartitionHand
+        // Avoid deadlocks by immediately scheduling a task for collecting dynamic filters because:
+        // * there can be task in other stage blocked waiting for the dynamic filters, or
+        // * connector split source for this stage might be blocked waiting the dynamic filters.
+        if (dynamicFilterService.isCollectingTaskNeeded(stageExecution.getStageId().getQueryId(), stageExecution.getFragment())) {
```
I'm not sure if scheduling is supposed to happen here. Could you please explain why is it necessary to schedule a task in `startLifespan`? Isn't the `schedule` method going to be called at least once before blocking?

sopel39 replied:
It's needed to collect dynamic filters for source stages with broadcast joins. 

> Isn't the schedule method going to be called at least once before blocking?

No. With phased scheduler some stages won't be scheduled before all dependent stages enumerate all splits. In this case there might be deadlock (without this extra task), because DF is collected by donwstream stage.

arhimondr replied:
This will effectively schedule a task on `StageScheduler` construction before `SqlQueryScheduler` is fully initialized and scheduling is started. 

At very least the state reporting will be inconsistent. The fact that there are some tasks running while the stage is still in a state `PLANNED` might be very non obvious and potentially error prone. Also it's not clear what is going to happen if a scheduled task fails or finishes before `SqlQueryScheduler` is fully initialized and all the listeners are attached. Even if it works today, it might easily break in the future, as it might not be expected for any tasks to exist before the `SqlQueryScheduler` is initialized and the stage is in state `SCHEDULING`. I would strongly recommend to avoid this route.

It generally feels like it should be responsibility of the `ExecutionSchedule` (`PhasedExecutionSchedule`) to determine what stages should be scheduled in what order. I was going to recommend making the `PhasedExecutionSchedule` aware of dynamic filters but it seems like you are already doing it. Based on the dynamic filter dependencies the `PhasedExecutionSchedule` can trigger scheduling for a source stage that contains broadcast join. Then the `SourcePartitionedScheduler` can rely on the existing mechanism on scheduling a single task when the split enumeration is blocked.

sopel39 replied:
I've added `io.trino.execution.scheduler.SourceScheduler#start` which is called from within `io.trino.execution.scheduler.SqlQueryScheduler.PipelinedDistributedStagesScheduler#schedule` for every `SourceScheduler`.

arhimondr commented on core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java:
```diff
@@ -490,10 +496,20 @@ private void collectDynamicFilters(StageId stageId, Optional<Set<DynamicFilterId
+        return ImmutableSet.copyOf(innerStageDynamicFilters);
+    }
+
+    private static Set<DynamicFilterId> getSourceDistributionStageProducedDynamicFilters(PlanFragment plan)
```
I assume this change is just a refactor. You just added an assertion that `producedDynamicFilters` are equal to `replicatedDynamicFilters`. And given `producedDynamicFilters` and `replicatedDynamicFilters` sets are equal, `producedDynamicFilters` x `consumedDymanicFilters` x `replicatedDynamicFilters` = `producedDynamicFilters` x `consumedDymanicFilters`.

Is my understanding correct?

arhimondr commented on core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java:
```diff
@@ -232,7 +232,13 @@ public boolean isCollectingTaskNeeded(QueryId queryId, PlanFragment plan)
+            return false;
+        }
+
+        return !getSourceDistributionStageProducedDynamicFilters(plan).isEmpty();
```
This changes the logic from `If there's any dynamic filters produced and consumed within a stage` to `If there's any dynamic filters produced by a stage`. Am i correct? Could you please explain why is it necessary? Is it needed to cover a mix of broadcast and partitioned joins?

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/scheduler/policy/PhasedExecutionSchedule.java:
```diff
@@ -410,19 +412,18 @@ private FragmentSubGraph processJoin(boolean replicated, PlanNode probe, PlanNod
-                // In case build source stage is blocked on full task buffer then join stage
-                // will be started automatically regardless od dependency. This is handled by
-                // unblockStagesWithFullOutputBuffer method.
+            if (fragments.get(currentFragmentId).getPartitioning().equals(SOURCE_DISTRIBUTION) && currentFragmentLazy) {
```
From what I understand it is to prevent a non source distributed stage running a replicated join from being executed lazily. Is it correct? I wonder why is it necessary.

arhimondr reviewed (commented):
It doesn't feel like I really understand the problem. Added some questions. 

It would be great if you could extract refactoring and improvements that are unrelated to have a commit that only does what's needed to address the problem. 

Additionally if you could elaborate more on a problem in the commit message it would be great. Maybe you can provide an example of a query that triggers a deadlock with a description of what join distributions are, where are the stage boundaries and how dynamic filters interact with each other.

sopel39 commented:
@arhimondr I've simplified this PR (removed refactor), keeping just needed parts. i cannot answer some outdated comments

> This changes the logic from If there's any dynamic filters produced and consumed within a stage to If there's any dynamic filters produced by a stage. Am i correct? Could you please explain why is it necessary? Is it needed to cover a mix of broadcast and partitioned joins?

I've changed to logic to create collecting task if there is any lazy DF produced by stage. This is needed because there might be consumers of that lazy DF outside of stage.

> From what I understand it is to prevent a non source distributed stage running a replicated join from being executed lazily. Is it correct? I wonder why is it necessary.

partitioned stages wont be covered by extra collecting task. Hence, if there are any lazy DFs produced by that stage, then stage needs to be scheduled

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -1542,6 +1542,8 @@ public void schedule()
         {
             checkState(started.compareAndSet(false, true), "already started");
 
+            stageSchedulers.values().forEach(StageScheduler::start);
```
nit: maybe move it inside the `try` block

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/scheduler/StageScheduler.java:
```diff
@@ -18,6 +18,8 @@
 public interface StageScheduler
         extends Closeable
 {
+    default void start() {}
```
Maybe add a short documentation, e.g.:

`Called by the query scheduler when the scheduling process begins. This method is called before the ExecutionSchedule takes a decision to schedule a stage but after the query scheduling has been fully initialized. Within this method the scheduler may decide to schedule tasks that are necessary for query execution to make progress. For example the the scheduler may decided to schedule a task without assigning any splits to unblock dynamic filter collection.`

sopel39 commented:
Thanks for review
